cert-manager 1.21: Cert-manager panics if Certificate.spec.renewal.policy is set to Disabled

**Describe the bug**:

If a Certificate sets the renewal policy field to "Disabled", the controller starts to panic and crashes.

```
panic: runtime error: invalid memory address or nil pointer dereference [recovered, repanicked]
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x27f366c]

goroutine 673 [running]:
sync.(*WaitGroup).Go.func1.1()
	sync/waitgroup.go:251 +0x45
panic({0x2cc9420?, 0x53119b0?})
	runtime/panic.go:860 +0x13a
github.com/cert-manager/cert-manager/pkg/controller/certificates/trigger.(*controllerWrapper).Register.NewTriggerPolicyChain.CurrentCertificateNearingExpiry.func1({0x3175fc10cc08, 0x3175fbe6c8c0, 0x3175fbfd0fc8, 0x0, 0x0})
	github.com/cert-manager/cert-manager@v0.0.0-00010101000000-000000000000/internal/controller/certificates/policies/checks.go:311 +0x38c
github.com/cert-manager/cert-manager/internal/controller/certificates/policies.Chain.Evaluate({0x3175fad301e0, 0x9, 0x0?}, {0x3175fc10cc08, 0x3175fbe6c8c0, 0x3175fbfd0fc8, 0x0, 0x0})
	github.com/cert-manager/cert-manager@v0.0.0-00010101000000-000000000000/internal/controller/certificates/policies/policies.go:61 +0x78
github.com/cert-manager/cert-manager/pkg/controller/certificates/trigger.(*controller).ProcessItem(0x3175fb37e460, {0x3484910, 0x3175fb09fd10}, {{0x3175fbfbd520, 0xd}, {0x3175fc1127e0, 0x19}})
	github.com/cert-manager/cert-manager@v0.0.0-00010101000000-000000000000/pkg/controller/certificates/trigger/trigger_controller.go:223 +0x603
github.com/cert-manager/cert-manager/pkg/controller.(*controller).worker.func1(0x3175fb802480, {{0x3175fbfbd520?, 0xf?}, {0x3175fc1127e0?, 0x0?}}, {{0x3491060?, 0x3175fab98000?}, 0x0?}, {0x3484910, 0x3175fb09fd10})
	github.com/cert-manager/cert-manager@v0.0.0-00010101000000-000000000000/pkg/controller/controller.go:147 +0x170
github.com/cert-manager/cert-manager/pkg/controller.(*controller).worker(0x3175fb802480, {0x3484910, 0x3175fb09fd10})
	github.com/cert-manager/cert-manager@v0.0.0-00010101000000-000000000000/pkg/controller/controller.go:164 +0xc5
github.com/cert-manager/cert-manager/pkg/controller.(*controller).Run.func1()
	github.com/cert-manager/cert-manager@v0.0.0-00010101000000-000000000000/pkg/controller/controller.go:107 +0x25
sync.(*WaitGroup).Go.func1()
	sync/waitgroup.go:258 +0x4a
created by sync.(*WaitGroup).Go in goroutine 174
	sync/waitgroup.go:238 +0x73
```

This seems to happen because the policy checking if a certificate should be renewed, assumes the response from `pki.RenewalTime` is non-nil:

https://github.com/cert-manager/cert-manager/blob/b8f325e36f49626ba72d7efbe138c01a5e661d96/internal/controller/certificates/policies/checks.go#L305-L310

But the function can return nil:

https://github.com/cert-manager/cert-manager/blob/b8f325e36f49626ba72d7efbe138c01a5e661d96/pkg/util/pki/renewaltime.go#L92-L93

And when this is assumed to be non-nil, it crashes:

https://github.com/cert-manager/cert-manager/blob/b8f325e36f49626ba72d7efbe138c01a5e661d96/internal/controller/certificates/policies/checks.go#L311

**Expected behaviour**:

The controller should not crash upon seeing certificates with renewal policy disabled.

**Steps to reproduce the bug**:

Create a selfsigned issuer:

```yaml
apiVersion: cert-manager.io/v1
kind: Issuer
metadata:
  name: selfsigned-issuer-v1
spec:
  selfSigned: {}
```

Create a certificate with renewal policy specified and set to Disabled:

```yaml
apiVersion: cert-manager.io/v1
kind: Certificate
metadata:
  name: root-ca-r1
spec:
  commonName: Root CA
  duration: 175200h
  isCA: true
  issuerRef:
    group: cert-manager.io
    kind: Issuer
    name: selfsigned-issuer-v1
  privateKey:
    algorithm: ECDSA
    rotationPolicy: Always
    size: 384
  renewBefore: 1h
  renewal:
    policy: Disabled
  secretName: root-ca-r1-secret
```

The controller now panics.

**Anything else we need to know?**:

**Environment details**:
- Kubernetes version: v1.36.1
- Cloud-provider/provisioner: KIND
- cert-manager version: v1.21.0
- Install method: `kubectl apply -f` from https://cert-manager.io/docs/installation/kubectl/

/kind bug

--- a/internal/controller/certificates/policies/checks.go
+++ b/internal/controller/certificates/policies/checks.go
@@ -273,8 +273,7 @@ func CurrentCertificateNearingExpiry(c clock.Clock) Func {
 
 		// If the certificate renewal policy is set to Disabled, do not renew.
 		if crt.Spec.Renewal != nil && crt.Spec.Renewal.Policy == cmapi.CertificateRenewalPolicyDisabled {
-			// renewal policy is set to disabled, do not renew
-			return RenewalDisabled, fmt.Sprintf("The certificate renewal policy is set to %s", cmapi.CertificateRenewalPolicyDisabled), true
+			return RenewalDisabled, fmt.Sprintf("The certificate renewal policy is set to %s", cmapi.CertificateRenewalPolicyDisabled), false
 		}
 
 		reason := Renewing

--- a/internal/controller/certificates/policies/checks.go
+++ b/internal/controller/certificates/policies/checks.go
@@ -271,6 +271,12 @@ func CurrentCertificateNearingExpiry(c clock.Clock) Func {
 		notAfter := metav1.NewTime(x509Cert.NotAfter)
 		crt := input.Certificate
 
+		// If the certificate renewal policy is set to Disabled, do not renew.
+		if crt.Spec.Renewal != nil && crt.Spec.Renewal.Policy == cmapi.CertificateRenewalPolicyDisabled {
+			// renewal policy is set to disabled, do not renew
+			return RenewalDisabled, fmt.Sprintf("The certificate renewal policy is set to %s", cmapi.CertificateRenewalPolicyDisabled), true
+		}
+
 		reason := Renewing
 		message := fmt.Sprintf("Renewing certificate as renewal was scheduled at %s", input.Certificate.Status.RenewalTime)
 

--- a/internal/controller/certificates/policies/checks.go
+++ b/internal/controller/certificates/policies/checks.go
@@ -273,7 +273,7 @@ func CurrentCertificateNearingExpiry(c clock.Clock) Func {
 
 		// If the certificate renewal policy is set to Disabled, do not renew.
 		if crt.Spec.Renewal != nil && crt.Spec.Renewal.Policy == cmapi.CertificateRenewalPolicyDisabled {
-			return RenewalDisabled, fmt.Sprintf("The certificate renewal policy is set to %s", cmapi.CertificateRenewalPolicyDisabled), false
+			return "", "", false
 		}
 
 		reason := Renewing

--- a/internal/controller/certificates/policies/checks_test.go
+++ b/internal/controller/certificates/policies/checks_test.go
@@ -543,6 +543,42 @@ func Test_NewTriggerPolicyChain(t *testing.T) {
 				},
 			},
 		},
+		"does not trigger renewal if renewal policy is disabled": {
+			certificate: &cmapi.Certificate{
+				Spec: cmapi.CertificateSpec{
+					CommonName: "example.com",
+					IssuerRef: cmmeta.IssuerReference{
+						Name:  "testissuer",
+						Kind:  "IssuerKind",
+						Group: "group.example.com",
+					},
+					Renewal: &cmapi.CertificateRenewal{
+						Policy: cmapi.CertificateRenewalPolicyDisabled,
+					},
+				},
+				Status: cmapi.CertificateStatus{
+					RenewalTime: nil,
+				},
+			},
+			secret: &corev1.Secret{
+				ObjectMeta: metav1.ObjectMeta{Name: "something",
+					Annotations: map[string]string{
+						cmapi.IssuerNameAnnotationKey:  "testissuer",
+						cmapi.IssuerKindAnnotationKey:  "IssuerKind",
+						cmapi.IssuerGroupAnnotationKey: "group.example.com",
+					},
+				},
+				Data: map[string][]byte{
+					corev1.TLSPrivateKeyKey: staticFixedPrivateKey,
+					corev1.TLSCertKey: testcrypto.MustCreateCertWithNotBeforeAfter(t, staticFixedPrivateKey,
+						&cmapi.Certificate{Spec: cmapi.CertificateSpec{CommonName: "example.com"}},
+						clock.Now().Add(time.Minute*-30),
+						// expires in 5 minutes time
+						clock.Now().Add(time.Minute*5),
+					),
+				},
+			},
+		},
 	}
 	policyChain := NewTriggerPolicyChain(clock)
 	for name, test := range tests {

--- a/internal/controller/certificates/policies/constants.go
+++ b/internal/controller/certificates/policies/constants.go
@@ -66,6 +66,9 @@ const (
 	// SecretManagedMetadataMismatch is a policy violation whereby the Secret is
 	// missing labels that should have been added by cert-manager
 	SecretManagedMetadataMismatch string = "SecretManagedMetadataMismatch"
+	// RenewalDisabled is a policy violation whereby the Certificate's
+	// Renewal Policy is set to RenewalDisabled.
+	RenewalDisabled string = "Disabled"
 
 	// AdditionalOutputFormatsMismatch is a policy violation whereby the
 	// Certificate's AdditionalOutputFormats is not reflected on the target

--- a/internal/controller/certificates/policies/constants.go
+++ b/internal/controller/certificates/policies/constants.go
@@ -67,7 +67,7 @@ const (
 	// missing labels that should have been added by cert-manager
 	SecretManagedMetadataMismatch string = "SecretManagedMetadataMismatch"
 	// RenewalDisabled is a policy violation whereby the Certificate's
-	// Renewal Policy is set to RenewalDisabled.
+	// Renewal Policy is set to Disabled.
 	RenewalDisabled string = "Disabled"
 
 	// AdditionalOutputFormatsMismatch is a policy violation whereby the

